samples/matter/common/src/util/finite_map.h

```diff
@@ -0,0 +1,121 @@
+/*
+ * Copyright (c) 2024 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-Nordic-5-Clause
+ */
+
+#pragma once
+
+#include <cstdint>
+#include <functional>
+#include <map>
+
+namespace Nrf
+{
+/*
+   FiniteMap template container allows to wrap mappings between uint16_t-type key and T-type value.
+   The size or the container is predefined at compilation time, therefore
+   the maximum number of stored elements must be well known. FiniteMap owns
+   inserted values, meaning that once the user inserts a value it will not longer
+   be valid outside of the container, i.e. FiniteMap cannot return stored object by copy.
+   FiniteMap's API offers basic operations like:
+     * inserting new values under provided key (Insert)
+     * erasing existing items under a given key (Erase) - this operation does not take care of the
+	   duplicated items,  meaning that the pointer under a given key is released unconditionally
+	   and the responsibility for the duplications is on the application side.
+     * retrieving values stored under provided key ([] operator)
+     * checking if the map contains a non-null value under given key (Contains)
+	 * retrieving a number of free slots available in the map (FreeSlots)
+     * iterating though stored item via publicly available mMap member
+     * retrieving the first free slot (GetFirstFreeSlot)
+	Prerequisites:
+     * T must have move semantics and bool()/==operators implemented
+*/
+template <typename T, std::size_t N> struct FiniteMap {
+	static constexpr uint16_t kInvalidKey{ N + 1 };
+	struct Item {
+		/* Initialize with invalid key (0 is a valid key) */
+		uint16_t key{ kInvalidKey };
+		T value;
+	};
+
+	bool Insert(uint16_t key, T &&value)
+	{
+		if (Contains(key)) {
+			/* The key with sane value already exists in the map, return prematurely. */
+			return false;
+		} else if (mElementsCount < N) {
+			mMap[key].key = key;
+			mMap[key].value = std::move(value);
+			mElementsCount++;
+		}
+		return true;
+	}
+
+	bool Erase(uint16_t key)
+	{
+		const auto &it = std::find_if(std::begin(mMap), std::end(mMap),
+					      [key](const Item &item) { return item.key == key; });
+		if (it != std::end(mMap)) {
+			it->value = T{};
+			it->key = kInvalidKey;
+			mElementsCount--;
+			return true;
+		}
+		return false;
+	}
+
+	/* Always use Constains() before using operator[]. */
+	T &operator[](uint16_t key)
+	{
+		static T dummyObject;
+		for (auto &it : mMap) {
+			if (key == it.key)
+				return it.value;
+		}
+		return dummyObject;
+	}
+
+	bool Contains(uint16_t key)
+	{
+		for (auto &it : mMap) {
+			if (key == it.key)
+				return true;
+		}
+		return false;
+	}
+
+	std::size_t FreeSlots() { return N - mElementsCount; }
+
+	uint16_t GetFirstFreeSlot()
+	{
+		uint16_t foundIndex = 0;
+		for (auto &it : mMap) {
+			if (kInvalidKey == it.key)
+				return foundIndex;
+			foundIndex++;
+		}
+		return -1;
```

Comment:
Is it done on purpose to underflow here, maybe it's better to return UINT16_MAX or kInvalidKey. Also I guess we could use `std::find_if`.